File: week4_paths2/1_dijkstra/dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <limits>
using namespace std;

/* using std::vector; */
/* using std::queue; */
/* using std::pair; */
/* using std::priority_queue; */
class Vertex
{
  public:
    int ID; //  zero-based index
    int dist; // distance estimate

    Vertex(int _ID, int _dist)
      :ID(_ID), dist(_dist)
    {
    }
};

class Compare_Dist
{
  public:
    bool operator () (Vertex & v1, Vertex & v2)
    {
      return v1.dist > v2.dist;  // Vertex of  smaller distance  has precedence.
    }
};

class Directed_Graph 
{
  private:
    vector<vector<int>> adj_list;
    vector<vector<int>> cost;
    int s, t;
    int int_max = numeric_limits<int>::max();
  
  public:
    Directed_Graph() {}
    void Data_Read()
    {
      int n, m;
      std::cin >> n >> m;
      adj_list.resize(n);
      cost.resize(n);

      for (int i = 0; i < m; i++) {
	int x, y, w;
	std::cin >> x >> y >> w;
	adj_list[x - 1].push_back(y - 1);   
	cost[x - 1].push_back(w);
	// convert to zero-based
	// such that index of array element corresponds to the vertex index.
      }
      std::cin >> s >> t;
      s--, t--;  // convert to zero-based index.
    }

    vector<int> Dijkstra(int source)
    {
      /* Dijkstra Algorithm to find the shortest path length with regard to minimum total weights.
	 Input:
	    source(int): The zero-based index of the source vertex.
	 Return:
	    Vector of distance from source vertex to all other vertices.
	 Remarks: Priority Queue based version.
	 Time Complexity: O((|V| + |E|) log|V|)
	 Space Complexity: O(|V|)
      */
      int num_of_vertices = adj_list.size();
      vector<int> dist(num_of_vertices, int_max);
      dist[source] = 0;
      priority_queue<Vertex , vector<Vertex>, Compare_Dist> H;
      H.push(Vertex(source, dist[source]));

      while(!H.empty())
      /* for(int i(0); i < num_of_vertices; i++) */
      {
	int u = H.top().ID; // return vertex with minimum dist.
	H.pop();
	for(int j(0); j < adj_list[u].size(); j ++)
	{
	  int v = adj_list[u][j];
	  int w = cost[u][j];
	  
	  // check for relaxation
	  if(dist[v] > (dist[u] + w))
	  {
	    dist[v] = dist[u] +w;
	    H.push(Vertex(v, dist[v]));
	  }
	}
      }

      return dist;
    }

    int process()
    {
      vector<int> dist = Dijkstra(s);
      return dist[t] != int_max ? dist[t] : -1;
    }
};


int main() {
  Directed_Graph graph;
  graph.Data_Read();
  cout << graph.process();

  /* int n, m; */
  /* std::cin >> n >> m; */
  /* vector<vector<int> > adj(n, vector<int>()); */
  /* vector<vector<int> > cost(n, vector<int>()); */
  /* for (int i = 0; i < m; i++) { */
  /*   int x, y, w; */
  /*   std::cin >> x >> y >> w; */
  /*   adj[x - 1].push_back(y - 1); */   
  /*   cost[x - 1].push_back(w); */
  /*   // convert to zero-based */
  /*   // such that index of array element corresponds to the vertex index. */
  /* } */
  /* int s, t; */
  /* std::cin >> s >> t; */
  /* s--, t--; */
  /* std::cout << distance(adj, cost, s, t); */
}

File: week4_paths2/1_dijkstra/dijkstra.py
#Uses python3

import sys
import queue
import heapq as pq

class Vertex:
    def __init__(self, _ID, _dist):
        self.ID = _ID
        self.dist = _dist

    def __lt__(self, other):
        return self.ID < other.ID # smaller ID has higher Priority

class Directed_Graph:
    def __init__(self):
        self.adj_list = None
        self.cost = None
        self.s = None
        self.t = None

    def Data_Read(self):
        input = sys.stdin.read()
        data = list(map(int, input.split()))
        n, m = data[0:2]
        data = data[2:]
        edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
        data = data[3 * m:]
        self.adj_list = [[] for _ in range(n)]
        self.cost = [[] for _ in range(n)]
        for ((a, b), w) in edges:
            self.adj_list[a - 1].append(b - 1)
            self.cost[a - 1].append(w)
        self.s, self.t = data[0] - 1, data[1] - 1

    def Dijkstra(self, source):
        ''' Dijkstra Algorithm to find the shortest path length with regard to minimum total weights.
            Input:
                source(int): The zero-based index of the source vertex.
            Return:
                Vector of distance from source vertex to all other vertices.
            Remarks: Priority Queue based version.
            Time Complexity: O((|V| + |E|) log|V|)
            Space Complexity: O(|V|)
        '''
        num_of_vertices = len(self.adj_list)
        dist = [float('inf')] * num_of_vertices
        dist[source] = 0
        H = []
        pq.heappush(H, Vertex(source, 0))
        while H:
            u = pq.heappop(H).ID
            for i in range(len(self.adj_list[u])):
                v = self.adj_list[u][i]
                w = self.cost[u][i]

                if dist[v] > dist[u] + w:
                    dist[v] = dist[u] + w
                    pq.heappush(H, Vertex(v, dist[v]))

        return dist

    def process(self):
        dist = self.Dijkstra(self.s)
        return dist[self.t] if dist[self.t] != float('inf') else -1


if __name__ == '__main__':
    graph = Directed_Graph()
    graph.Data_Read()
    print(graph.process())
    # input = sys.stdin.read()
    # data = list(map(int, input.split()))
    # n, m = data[0:2]
    # data = data[2:]
    # edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    # data = data[3 * m:]
    # adj = [[] for _ in range(n)]
    # cost = [[] for _ in range(n)]
    # for ((a, b), w) in edges:
    #     adj[a - 1].append(b - 1)
    #     cost[a - 1].append(w)
    # s, t = data[0] - 1, data[1] - 1
    # print(distance(adj, cost, s, t))
